tree/dataframe/inc/ROOT/RDFHelpers.hxx

```diff
@@ -181,6 +184,110 @@ RNode AsRNode(NodeType node)
 // clang-format on
 void RunGraphs(std::vector<RResultHandle> handles);
 
+
+// clang-format off
+/// RDF progress helper.
+/// This class provides callback functions to RDataFrame, which record event statistics
+/// and print them to the terminal every second.
+/// ProgressHelper::operator()(unsigned int, T&) is thread safe, and can be used as callback in MT mode.
+// clang-format on
+class ProgressHelper {
+public:
+
+  /// Create a progress helper.
+  /// \param increment RDF callbacks are called every `n` events. Pass this `n` here.
+  /// \param maxEvent If known, pass the number of events in the dataset. This will allow for computing completion
+  /// statistics.
+  /// \param progressBarWidth Number of characters the progress bar will occupy.
+  /// \param useShellColours Use shell colour codes to colour the output.
+  ProgressHelper(std::size_t increment,
+      std::size_t maxEvent,
+      unsigned int progressBarWidth = 40,
+      bool useShellColours = true) :
+  fMaxEvents{maxEvent},
+  fIncrement{increment},
+  fBarWidth{progressBarWidth},
+  fUseShellColours{useShellColours}
+  { }
+
+  ~ProgressHelper() = default;
+
+  // clang-format off
+  /// Thread-safe callback for RDataFrame.
+  /// It will record elapsed times and event statistics, and print a progress bar every second.
+  /// \param slot Ignored.
+  /// \param value Ignored.
+  // clang-format on
+  template<typename T>
+  void operator()(unsigned int /*slot*/, T& value) { operator()(value); }
+
+  // clang-format off
+  /// Thread-safe callback for RDataFrame.
+  /// It will record elapsed times and event statistics, and print a progress bar every second.
+  /// \param value Ignored.
+  // clang-format on
+  template<typename T>
+  void operator()(T& /*value*/) {
+    using namespace std::chrono;
+    // ***************************************************
+    // Warning: Here, everything needs to be thread safe:
+    // ***************************************************
+    fProcessedEvents += fIncrement;
+
+    // We only print every 1s, and only one thread does it. Try to take the lock:
+    if (duration_cast<seconds>(system_clock::now() - fLastPrintTime).count() == 0 || !fPrintMutex.try_lock())
+      return;
+
+    // ***************************************************
+    // Protected by lock from here:
+    // ***************************************************
+    std::lock_guard<std::mutex> lockGuard(fPrintMutex, std::adopt_lock_t{});
+
+    std::size_t eventCount;
+    seconds elapsedSeconds;
+    std::tie(eventCount, elapsedSeconds) = RecordEvtCountAndTime();
+
+    std::cout << "\r";
```

Comment:
What happens on Windows?

Reply:
As a first step: Don't use it on windows.
Second step: I don't know. What should happen?

Reply:
What should happen? A nice progress bra is displayed, or at least the terminal isn't flooded with new lines with some number of `=` :-)

Reply:
Yes, yes, sure. Let me rephrase the question:
What do we have to do to get something working on windows?
I don't know how their terminals work, and I don't have the time to look it up.

And why do you want a
> bra
to be displayed? Can we settle for a bar as a first step? 😄